### PR TITLE
#1387 Update Tooltips in textFields

### PR DESCRIPTION
**Pull request by nabilIbm**
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx [hunk 1]
```diff
@@ -109,25 +110,26 @@ class TextfieldControl extends React.Component {
...
+			let content = textInput;
 			let disabled = true;
 			if (value && this.props.state !== STATES.DISABLED) {
 				disabled = false;
 			}
-			const tooltip = (
-				<div className="properties-tooltips">
-					{String(value)}
-				</div>
+			if (this.props.control.icon) {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx:118]: Why add the icon?  This shouldn't be added for textfields.

**Review by matthoward366 — CHANGES_REQUESTED** [review]
I don't see tooltips working anymore for textfields in tables.

File: canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx [hunk 2]
```diff
@@ -109,25 +110,26 @@ class TextfieldControl extends React.Component {
...
+						<div className={"properties-field-type-icon"}>
+							<Icon type={this.props.control.icon} />
+						</div>
+						{textInput}
+					</div>);
+			}
+			display = (
+				<TruncatedContentTooltip
+					content={content}
+					disabled={disabled}
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx:130]: You're missing the `tooltipText` that is marked as required in this component
image

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 3]
```diff
@@ -107,7 +107,7 @@ class ToolTip extends React.Component {
 		const canDisplayFullText = this.canDisplayFullText(this.triggerRef);
 		const showToolTip = (
 			// show tooltip if not disabled and showToolTipIfTruncated is false
-			(!this.props.disable && !this.props.showToolTipIfTruncated) ||
+			(!this.props.disable && !this.props.showToolTipIfTruncated && !canDisplayFullText) ||
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:110]: This is the same as line 112. This doesn't seem right.  Why was this changed?

Comment (nabilIbm) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:110]: I added  it to have the expected label's behavior the same as textfiled (otherwise tooltip will show in ALL header values )
I'll  update the function below then `canDisplayFullText` , see  what needs to be changed to handle both (label & textfiled)

Comment (matthoward366) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:110]: For the table header it might be caused by using `Tooltip` instead of TruncatedContentTooltip

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 4]
```diff
@@ -119,7 +119,8 @@ class ToolTip extends React.Component {
 	// (scrollWidth) value is equal to the minimum width the element would require
 	//  in order to fit all the content in the viewport without using a horizontal scrollbar
 	canDisplayFullText(elem) {
-		if (elem) {
+		const input = elem.getElementsByTagName("input")[0];
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:122]: This doesn't seem like the right place for this.  It's too specific to `input` elements and the tooltip component is more generic.  I would probably pass in a new property into the tooltip component to check.  Something like `truncatedElement`.  Then pass the `input` element into that.  It would be optional and fall back to this.props.children if not set.

**Review by matthoward366 — CHANGES_REQUESTED** [review]

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx (hunk 3 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:110]: I tried it in first but i got all textfield gone from the UI
image

Comment (nabilIbm) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:110]: What i was trying to say in here,  is im using `TruncatedContentTooltip` not `Tooltip` (if this was your comment) and yet labels didn't work 
image

Comment (matthoward366) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:110]: You should update your PR with the correct code then figure out why it doesn't work correctly.

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 5]
```diff
@@ -30,6 +30,7 @@ class ToolTip extends React.Component {
 		};
 
 		this.pendingTooltip = null;
+		this.truncatedElement = null;
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:33]: This should be a property passed in by the TextField control.

Comment (matthoward366) [reply canvas_modules/common-canvas/src/tooltip/tooltip.jsx:33]: Then that would be used instead of `elem` in canDisplayFullText.

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 6]
```diff
@@ -331,6 +332,8 @@ class ToolTip extends React.Component {
 	}
 
 	render() {
+		const inputTruncated = this.triggerRef && this.triggerRef.getElementsByTagName("input")[0];
+		this.truncatedElement = !(inputTruncated && inputTruncated.scrollWidth > inputTruncated.clientWidth);
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:336]: Hi @nabilIbm, this code right here is what's calculating if the Textfield should be truncated or not. This logic should be moved to `TruncatedContentTooltip` instead. Then no changes are required in `Tooltip`. Inside `TruncatedContentTooltip`, if you determine that the `content` needs to be truncated, you will return the `<Tooltip>`, otherwise, just return the `content`

File: canvas_modules/common-canvas/src/tooltip/tooltip.jsx [hunk 7]
```diff
@@ -440,14 +420,16 @@ ToolTip.propTypes = {
 	disable: PropTypes.bool, // Tooltip will not show if disabled
 	showToolTipIfTruncated: PropTypes.bool, // Set to true to only display tooltip if full text does not fit in displayable width
 	delay: PropTypes.number,
-	showToolTipOnClick: PropTypes.bool
+	showToolTipOnClick: PropTypes.bool,
+	truncatElem: PropTypes.object,
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/tooltip/tooltip.jsx:424]: This shouldn't be needed since this is handled in TruncatedTooltip.

File: canvas_modules/common-canvas/src/common-properties/components/truncated-content-tooltip/truncated-content-tooltip.jsx [hunk 8]
```diff
@@ -22,8 +22,27 @@ import { has } from "lodash";
 
 // Reusable component to show tooltip if the content is truncated
 export default class TruncatedContentTooltip extends React.Component {
+	// Return true if the string can be displayed in the available space
+	// Return false if the string is truncated and ellipsis is shown on the UI
+	// (offsetWidth) is a measurement in pixels of the element's CSS width, including any borders, padding, and vertical scrollbars
+	// (scrollWidth) value is equal to the minimum width the element would require
+	//  in order to fit all the content in the viewport without using a horizontal scrollbar
+	canDisplayFullText(elem) {
+		if (elem) {
```
Comment (matthoward366) [inline canvas_modules/common-canvas/src/common-properties/components/truncated-content-tooltip/truncated-content-tooltip.jsx:31]: I would expect this to be `elem` or `truncatElem`.  Then textfield passes in `truncatElem`

**Review by matthoward366 — CHANGES_REQUESTED** [review]

**Comment by caritaou** [comment]
Hey @matthoward366, with the changes from this PR, should we also export `TruncatedContentTooltip` in https://github.com/elyra-ai/canvas/blob/main/canvas_modules/common-canvas/src/index.js? Some applications using `Tooltip` pass in `showToolTipIfTruncated`, so they will need to use `TruncatedContentTooltip` instead

**Comment by matthoward366** [comment]
@caritaou Forgot about that.  Why even have `TruncatedContentTooltip`?  It probably makes more sense to adjust Tooltip to deal with the css needed from `TruncatedContentTooltip`.

**Comment by nabilIbm** [comment]
Do we just keep `Tooltip` in both textfiled (text field)  & virtualized-table (for labels) and adjust them in css to inherit `TruncatedContentTooltip` behavior ?

**Comment by matthoward366** [comment]
You'd `TruncatedContentTooltip` completely and correct by updating Tooltip to use the css from TruncatedContentTooltip when `showToolTipIfTruncated` is passed in.

**Comment by matthoward366** [comment]
Replaced by https://github.com/elyra-ai/canvas/pull/1409
